src/wp-includes/template.php

```diff
@@ -823,3 +823,119 @@ function load_template( $_template_file, $load_once = true, $args = array() ) {
+		 * wp_cache_close(). If the buffers are not all flushed before wp_cache_close() is closed, then some output buffer
+		 * handlers (e.g. for caching plugins) may fail to be able to store the page output in the object cache.
+		 * See <https://github.com/WordPress/performance/pull/1317#issuecomment-2271955356>.
+		 */
+		PHP_OUTPUT_HANDLER_STDFLAGS ^ PHP_OUTPUT_HANDLER_FLUSHABLE
+	);
+}
+
+/**
+ * Finalizes the output buffer for the included template.
+ *
+ * @since 6.9.0
+ *
+ * @see wp_start_template_output_buffer()
+ *
+ * @param string $output Output buffer.
+ * @param int    $phase  Phase.
+ * @return string Finalized output buffer.
+ */
+function wp_finalize_template_output_buffer( string $output, int $phase ): string {
+	// When the output is being cleaned (e.g. pending template is replaced with error page), do not send it through the filter.
+	if ( ( $phase & PHP_OUTPUT_HANDLER_CLEAN ) !== 0 ) {
+		return $output;
+	}
+
+	// Detect if the response is an HTML content type.
+	$is_html_content_type = false;
+	$headers_list         = array_merge(
+		array( 'Content-Type: ' . ini_get( 'default_mimetype' ) ),
+		headers_list()
+	);
+	foreach ( $headers_list as $header ) {
+		$header_parts = preg_split( '/\s*[:;]\s*/', strtolower( $header ) );
+		if (
+			is_array( $header_parts ) &&
+			count( $header_parts ) >= 2 &&
+			'content-type' === $header_parts[0]
+		) {
+			$is_html_content_type = in_array( $header_parts[1], array( 'text/html', 'application/xhtml+xml' ), true );
+		}
+	}
+
+	// If the content type is HTML, require that there be at least one tag.
+	if ( $is_html_content_type ) {
+		$is_html_content_type = ( new WP_HTML_Tag_Processor( $output ) )->next_tag();
+	}
+
+	$filtered_output = $output;
+	if ( $is_html_content_type ) {
+		/**
+		 * Filters the HTML template output buffer prior to sending to the client.
+		 *
+		 * This filter only applies to HTML output, so it is safe for processing with the HTML API.
+		 * Notice: It is highly discouraged to use regular expressions to do any kind of replacement on the output.
+		 *
+		 * @since 6.9.0
+		 *
+		 * @param string $output Output buffer HTML.
+		 * @return string Filtered output buffer HTML.
+		 */
+		$filtered_output = (string) apply_filters( 'wp_template_output_buffer_html', $filtered_output );
+	}
+
+	/**
+	 * Filters the template output buffer prior to sending to the client.
+	 *
+	 * This filter applies to HTML and non-HTML template output alike. For example, a "template" could return JSON. Use
+	 * the {@see 'wp_template_output_buffer_html'}  filter specifically for filtering HTML output, which is the normal
+	 * case. Notice: It is highly discouraged to use regular expressions to do any kind of replacement on the output.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @param string $output Output buffer.
+	 * @return string Filtered output buffer.
+	 */
+	$filtered_output = (string) apply_filters( 'wp_template_output_buffer', $filtered_output );
```

Comment:
I think that this can be combined with the `wp_template_output_buffer_html` filter and  `$is_html_content_type` could be passed as an additional variable.  It also might be useful to include `$output`. 

My thinking is that running this through less hooks is going to make it easier for extenders to know where to hook into and introduce a bit less complexity. Additionally, `$output` being available here makes sense to me since it's in the end of function hook below

Reply:
My thinking for `wp_template_output_buffer_html` was that it would be a convenience for authors to have a filter which would predictably always be sent HTML, and this could always be sent into the HTML API. Otherwise, they'd need to always sniff the type of the output themselves (or look at an additional filter param). So having `wp_template_output_buffer_html` and `wp_template_output_buffer` filters is similar to how we have `render_block_{$block_name}` and `render_block` filters.

Reply:
I like how this introduces the possibility for a filter to undo any of the other changes that had been done to the `$filtered_output`.